File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int number_lenght(long number);
int luhns_alg(long cardNumber);
int get_first_n_numbers(long number, int n_numbers);

int main(void)
{
    long card = get_long("Number: ");
    int number_of_digits = number_lenght(card);
    int first = get_first_n_numbers(card, 1);
    string card_flag = "INVALID";

    if (number_of_digits >= 13 && number_of_digits <= 16)
    {
        int verify = luhns_alg(card);
        if ((verify % 10) != 0)
        {
            card_flag = "INVALID";
        }
        else if ((number_of_digits == 13 || number_of_digits == 16) &&
                 get_first_n_numbers(card, 1) == 4)
        {
            card_flag = "VISA";
        }
        else if ((number_of_digits == 16) &&
                 (get_first_n_numbers(card, 2) >= 51 && get_first_n_numbers(card, 2) <= 55))
        {
            card_flag = "MASTERCARD";
        }
        else if ((number_of_digits == 15) &&
                 (get_first_n_numbers(card, 2) == 34 || get_first_n_numbers(card, 2) == 37))
        {
            card_flag = "AMEX";
        }
    }

    printf("%s\n", card_flag);
}

int luhns_alg(long cardNumber)
{
    int sum_second_to_last = 0;
    int every_other = 0;
    int sum_of_others = 0;

    while (cardNumber > 0)
    {
        int last_digit = cardNumber % 10;
        if (every_other == 1)
        {
            int product = last_digit *= 2;
            if ((product / 10) > 0) // if the product have two digits
            {
                int first = product % 10;
                int last = product / 10;
                sum_second_to_last += first + last;
            }
            else
            {
                sum_second_to_last += product;
            }
            every_other = 0;
        }
        else
        {
            sum_of_others += last_digit;
            every_other = 1;
        }
        cardNumber /= 10;
    }
    return (sum_second_to_last + sum_of_others);
}

int number_lenght(long number)
{
    int number_of_digits = 0;

    while (number > 0)
    {
        number_of_digits++;
        number /= 10;
    }

    return number_of_digits;
}

int get_first_n_numbers(long number, int n_numbers)
{
    int x;
    if (n_numbers == 2)
    {
        x = 100;
    }
    else if (n_numbers == 1)
    {
        x = 10;
    }
    else
    {
        return 0;
    }

    while (number >= x)
    {
        number /= 10;
    }
    return number;
}

File: pset6/sentimental-credit/credit.py
def main():
    # Prompt for a credit card number
    while True:
        try:
            number = int(input("Number: "))
            break
        except ValueError:
            pass

    # Check if number is a valid credit card number
    flag = card_flag(number)

    if flag and luhn(number):
        # Print credit card flag name
        print(flag)
    else:
        print("INVALID")


# Algorithm used to validate credit card numbers
def luhn(card):
    card = str(card)  # Converts argument to str
    stl_digits = []  # Stores every-other-digit from second to last
    total_sum = 0

    # Iterate through every digit in card_number, starting from the last
    i = len(card) - 1  # Set i to last index of card_numbe
    while i >= 0:
        total_sum += int(card[i])
        if i > 0:
            stl_digits.append(card[i - 1])
        i -= 2

    # Multiply each stl_digits by 2 and sum the products
    for n in stl_digits:
        n = int(n) * 2
        # If the product have two digits
        if n > 9:
            # Sum the first and second digit
            n = int(n / 10) + int(n % 10)
        # Add number to total_sum
        total_sum += n

    # Check if total_sum's last digit is 0
    if total_sum % 10 == 0:
        return True
    else:
        return False


def card_flag(card):
    # Check if the number of digits is between 13 and 16 (inclusive)
    card = str(card)  # Convert argument to str
    n_digits = len(card)

    if (12 < n_digits < 17):
        # Check if number matches with some card flag (AMEX, MASTERCARD, VISA)
        if n_digits in [13, 16] and int(card[0]) == 4:
            return "VISA"
        elif n_digits == 15 and int(card[0] + card[1]) in [34, 37]:
            return "AMEX"
        elif n_digits == 16 and int(card[0] + card[1]) in [51, 52, 53, 54, 55]:
            return "MASTERCARD"
        else:
            return False


if __name__ == "__main__":
    main()
